### ISSUE TITLE
Fixed: Add sameSite attribute to Backdrop tableDrag cookie.

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4718]
When visiting a page with tabledrag, Firefox nags into the console:

`Cookie “Backdrop.tableDrag.showWeight” will be soon rejected because it has the “sameSite” attribute set to “none” or an invalid value, without the “secure” attribute.
`
More information about that nagging: https://developer.mozilla.org/de/docs/Web/HTTP/Headers/Set-Cookie/SameSite

Backdrop uses [jquery.cookie.js](https://github.com/backdrop/backdrop/blob/1.x/core/misc/jquery.cookie.js)

Actually using a cookie for tabledrag is not necessary it just toggles display stuff via js. I'd recommend to use [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) instead.

This is a related [Drupal 8 issue](https://www.drupal.org/project/drupal/issues/1569648), where they switched to localStorage as well.

**Comment by laryn** [comment]
Related?
https://www.zoocha.com/news/how-fix-chrome-80-cookies-issue-drupal

**Comment by indigoxela** [comment]
> Related?

Yep, seems to be related.

I have to admit, that I'm really no expert regarding that topic.

Another thing I just figured out: Backdrop ships with jquery.cookie.js version 1.0 (2006), but that's pretty outdated:
https://github.com/js-cookie/js-cookie

It seems that sameSite attribute support has been added in 2016: https://github.com/js-cookie/js-cookie/issues/276

We should really update jquery.cookie.js.

**Comment by quicksketch** [comment]
Considering `js-cookie` and `jquery.cookie.js` were written by the same author, they helpfully provided a backwards-compatible approach that we might be able to use: https://github.com/js-cookie/js-cookie/tree/v1.5.1#migrating-from-jquery-cookie

```
$.cookie('name', 'value') === Cookies.set('name', 'value')
$.cookie('name') === Cookies.get('name')
$.removeCookie('name') === Cookies.remove('name')
$.cookie() === Cookies.get()
```

Though I'm not sure if that's really a good idea. Backdrop core's usage of the library is pretty minimal though, only in Tabledrag, Contact module, and Comment module.

**Comment by indigoxela** [comment]
Maybe this [D8 issue](https://www.drupal.org/project/drupal/issues/2550717) is useful here?

**Comment by indigoxela** [comment]
Regarding tabledrag.js and the annoying cookie console log... Heck, why don't we simply switch to localstorage? There's actually no need for a cookie (which always gets sent via HTTP), because the switching happens in js in the client, anyway. Or am I missing something?

Examples:
```
localStorage.getItem('Backdrop.tableDrag.showWeight');
localStorage.setItem('Backdrop.tableDrag.showWeight', '1');
```
... as simple as that. :grin:

**Comment by indigoxela** [comment]
Coming back here after some time. A [PR](https://github.com/backdrop/backdrop/pull/3869) is available for testing and review.

How to test:

- Go to a page with tabledrag, like /admin/structure/types/manage/post/fields or /admin/structure/menu/manage/main-menu
- Toggle the "Show row weights" setting
- Check if toggle on/off works as expected

If you're more curious what happens, open the browser dev tools an check Web storage / Local storage for the toggling value of Backdrop.tableDrag.showWeight.
Note that a cookie might still exist, if you test locally, but it's not in use anymore and doesn't cause anymore console nagging.

**Comment by docwilmot** [comment]
I believe this looks good. If it works for the Firefox users I'd RTBC it, but cant test without FF.

**Comment by herbdool** [comment]
I've tested in FF. Code looks good.

Should we have follow-up issues for changing the cookie library in Contact and Comment as well?

**Comment by klonos** [comment]
Although the PR by @indigoxela works fine and it is an improvement, I'd be more comfortable if we followed Drupal changes more closely where possible, because:
- Drupal has a bigger user base, so things get more real-life testing
- If security/performance fixes come out for Drupal, it'll be harder for us the more we have deviated.

Also, from https://www.drupal.org/project/drupal/issues/1569648 that you have linked @indigoxela 

> Also using localStorage mean we can listen to events and react in background tabs. Open a "manage fields" page for some content type, then open another tab/window in the same or another "manage fields" page. Click "Show row weight" in one tab and look at the other one, the change propagated.

I have tested the above in the now "Manage displays" tab in D8/9, and it does indeed work w/o needing to refresh the page. That doesn't happen in the PR that's been RTBCed here.

With all respect to your solution @indigoxela, here's an alternative PR for consideration, which mimics the changes that went into Drupal: https://github.com/backdrop/backdrop/pull/3870

**Comment by klonos** [comment]
> Should we have follow-up issues for changing the cookie library in Contact and Comment as well?

I think so, yes 👍🏼 

I also think we need a follow-up to backport other fixes that have been committed to tabledrag.js over the past 8-9 years. I'll create those follow-ups once this is merged.

**Comment by indigoxela** [comment]
I've updated my PR to not set a value by default, which is more closely to what D8 did - and this is the part that makes sense to me. :wink: 

As we now have a competing PR here, IMO it makes sense to go back to "needs testing/review".

@klonos I left a hint on your PR re the deprecated "proxy" function.

**Comment by klonos** [comment]
Thanks @indigoxela ...I missed those 🤦🏼 

Fixing...

**Comment by herbdool** [comment]
I don't want to have to review both, should we close one? They're both pretty close. I can see the value of properties propagating between browser tabs though.

**Comment by klonos** [comment]
> I can see the value of properties propagating between browser tabs though.

Yes, I thought that that would be a nice UX improvement. People that toggle row weights in one table will most likely want that setting to propagate instead of going through all their tabs and doing it manually.

**Comment by indigoxela** [comment]
> I don't want to have to review both, should we close one?

Sure, probably makes sense.

> I can see the value of properties propagating between browser tabs though.

I don't. That and the quirky code with deprecated functions prevented me from "blindly" using the Drupal solution. Usually I prefer to only fix what hast been requested, keeping PRs uncluttered and easier to review.

Anyway, to not block this issue, I'll close mine.

**Comment by klonos** [comment]
@indigoxela please see my comment in the PR re `.proxy()`. It is deprecated only in jQuery v3.3, and we're still on v1 in Backdrop.

I've corrected the `.bind()` I've accidentally introduced into a `.on()` instead. So my PR should be ready for review again.

**Comment by indigoxela** [comment]
> It is deprecated only in jQuery v3.3, and we're still on v1 in Backdrop. I would still fix it, but it's being used in many other places in core,

@klonos I'm aware of that but this is an addition, and I don't think we should use deprecated functions in newly added code. That will just cause us more work, when we have to replace them later on. Why not doing it right when adding it?

**Comment by klonos** [comment]
> ...I don't think we should use deprecated functions in newly added code. That will just cause us more work

Fair enough. I tried to find out how `.proxy()` is to be alternatively implemented:

- The notice in https://api.jquery.com/jQuery.proxy says:
  > Note: This API has been deprecated in jQuery 3.3; please use the native `Function.prototype.bind` method instead.
  But isn't `.bind()` already deprecated? I'm confused 😕 
- The [change record for v3.3](https://blog.jquery.com/2018/01/19/jquery-3-3-0-a-fragrant-bouquet-of-deprecations-and-is-that-a-new-feature/) mentions:
  > ... While most of these will be removed in jQuery 4.0, it’s worth noting that we do not consider the deprecation of a method to mean that it will be removed; it means that we encourage the use of alternatives.
  > - ...
  > - jQuery.proxy (not slated for removal) ([#2958](https://github.com/jquery/jquery/issues/2958), [3a8e4474](https://github.com/jquery/jquery/commit/3a8e44745c014871bc56e94d91e57c45ae4be208))
  > - ...

So unless someone points me to documentation that explains how to replace `.proxy()`, I'm afraid I don't know what needs to be done. Perhaps someone with better knowledge of JS/jQuery can file a PR against mine, or a new PR based on mine 🤷🏼

**Comment by indigoxela** [comment]
> Perhaps someone with better knowledge of JS/jQuery can file a PR against mine, or a new PR based on mine

Or, alternatively, re-open mine. Sorry for suggesting that, but my PR does not introduce deprecated functions, does not introduce features nobody requested, does not introduce global js variables, simplifies code instead of complicating it.

But somebody else has to decide. :stuck_out_tongue_winking_eye:

**Comment by klonos** [comment]
Re-iterating what I said earlier:
> - Drupal has a bigger user base, so things get more real-life testing
> - If security/performance fixes come out for Drupal, it'll be harder for us the more we have deviated.

Baking our own code for the same functionality makes it more difficult to apply security fixes, bug fixes, and new features added in the respective Drupal code.

> But somebody else has to decide. 😜

I agree with that ^^ 👍🏼

**Comment by indigoxela** [comment]
> Baking our own code for the same functionality makes it more difficult to apply security fixes...

I get that and agree. However, Backdrop is different in many other places. Not every Drupal code change is "golden" and this tabledrag change is a clear example for that. :grin: It seems odd to me to prefer that, just because it comes from Drupal.

**Comment by herbdool** [comment]
I didn't look at the code too closely. I agree that we should not introduce deprecated functions if we can help it.

**Comment by argiepiano** [comment]
>Fair enough. I tried to find out how .proxy() is to be alternatively implemented:
...
>> Note: This API has been deprecated in jQuery 3.3; please use the native  Function.prototype.bind method instead.
But isn't .bind() already deprecated? I'm confused 😕

I have no opinion about this discussion, and I'm not familiar with the competing PRs either. 

Just wanted to point out that `jQuery.proxy()` CAN be replaced by the **native** `Function.prototype.bind()`. This `bind()` is a native JS method (not jQuery) that has similar use as jQuery.proxy(), and is not deprecated. 

As for `jQuery.bind()`, the [docs](https://api.jquery.com/bind/) say that it should be replaced by `jQuery.on()`.

**Comment by docwilmot** [comment]
I'd just like to say, and I've said this before, I think it is best if a person is looking into an existing issue, especially if they have an existing PR under discussion, that others make suggestions to that person, rather than taking over the issue with a new PR and new discussion. 

The conversation could have been a lot simpler: "hey @indigoxela Drupal's solution allows this to update through multiple tabs, can you incorporate that code?" Done. Now instead we have two PRs and an argument, instead of a cooperative solution.

**Comment by herbdool** [comment]
I agree with @docwilmot. Though I may have contributed to the confusion here. @klonos it would be better to provide snippets instead of a whole different pr.

@indigoxela what do you think of adapting your PR? Or prefer to have a follow-up for the improvements?

**Comment by indigoxela** [comment]
> @indigoxela what do you think of adapting your PR?

Sure, let's not get derailed here. :wink: PR has been updated.

**Comment by herbdool** [comment]
@indigoxela I've reviewed and tested your PR. Looks good.

**Comment by klonos** [comment]
Sorry guys. I would have made code snippet suggestions (you know I always do), but the code seemed totally different, so that wasn't possible/easy.

I still think that there's value in keeping code parity with Drupal commits where possible (as well as feature parity), as I think that it makes back/cross-ports easier even for less experienced developers, but happy to close my PR if there's no consensus re that. If the toggle propagation to all active tabs is requested by users in the future, I personally won't be able to contribute - we'd need someone with JS expertise to do that for us in a custom way (as opposed to the ready solution we have from the Drupal commit).

Having said all that, I understand the concerns re introducing deprecated functions, and I have now removed usage of `.proxy()` from my PR.

**Comment by indigoxela** [comment]
@klonos This throws the issue back to "needs more feedback". You seem to insist in your solution, which undermines the current status - it was already RTBC. I don't see any chance to solve this conflict without the PMC.

No, I won't close my PR. You won't close yours, I assume. That means that an easy fix gets stuck.

**Comment by klonos** [comment]
I don't think that this issue needs to go to the PMC. We have 4 active core committers, so they can decide what is the best way forward.

We have a unique solution, which works fine, and we have an alternative, which also works fine. There were concerns with my solution (which is basically a back/cross-port of the Drupal solution), which I have now addressed (no more usage of deprecated jQuery functions). Both are valid the way I see it - we just need to let the core committers decide which is the most appropriate way to solve this. This is nothing new, so no dramas.

**Comment by quicksketch** [comment]
I read both PRs and personally I find @indigoxela's approach to be easier to read and understand. Despite the benefit of matching Drupal code, I would prefer to go with @indigoxela's approach because it may be easier for our own community to maintain it independently.

In testing, I found that both approaches immediately update the columns across tabs (which I find unexpected, but in some cases could actually be useful). Since it's the same in both, that probably doesn't need to be part of the consideration as to which we should use.

**Comment by klonos** [comment]
Thanks @quicksketch 👍🏼 ...I've closed my PR then.

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/3869. Thanks everyone for the work here. Major props to @indigoxela for her solution and thanks to @klonos for being gracious in coming to an agreement. And thanks to reviewers @herbdool, @argiepiano, and @docwilmot. I'm really glad when our community can work out these minor conflicts. Thank you all! :bow:
